api/v1/controllers/collectors.js

```diff
@@ -320,18 +320,75 @@ function heartbeat(req, res, next) {
 } // heartbeat
 
 /**
- * Change collector status to Running. Invalid if the collector's status is
- * not Stopped.
+ * Creates a collector if name not found, with the user as the sole writer.
+ * Change collector status to Running and returns a new collector token.
+ * Reject if the user is not among the writers.
+ * Invalid if the collector's status is not Stopped.
  *
  * @param {IncomingMessage} req - The request object
  * @param {ServerResponse} res - The response object
  * @param {Function} next - The next middleware function in the stack
  */
 function startCollector(req, res, next) {
-  req.swagger.params.queryBody = {
-    value: { status: 'Running' },
-  };
-  doPatch(req, res, next, helper);
+  const toPost = req.swagger.params.queryBody.value;
+  toPost.status = 'Running';
+  let collector;
+
+  // returns null if no collector found
+  return helper.model.findOne({ where: { name: toPost.name } })
+  .then((returnedCollector) => {
+    if (!returnedCollector) {
+
+      // create collector, find user
+      const username = jwtUtil.getTokenDetailsFromRequest(req).username;
+      return Promise.all([
+        helper.model.create(toPost),
+        userProps.model.findOne({
+          where: {
+            name: username,
+          },
+        }),
+      ])
+      .then((results) => {
+        collector = results[0];
+        return collector.addWriter(results[1]);
+      })
+      .then(() => {
+
+        /*
+         * When a collector registers itself with Refocus, Refocus sends back a
+         * special token for that collector to use for all subsequent heartbeats.
+         */
+        collector.dataValues.token = jwtUtil
+          .createToken(toPost.name, toPost.name);
+        return res.status(httpStatus.OK)
+          .json(u.responsify(collector, helper, req.method));
+      });
+    }
+
+    // found collector
+    if (!returnedCollector.registered) {
+      throw new apiErrors.ForbiddenError({ explanation:
+        'Cannot start--this collector is not registered.',
+      });
+    }
+
+    if (returnedCollector.status === 'Running') {
+      throw new apiErrors.ForbiddenError({ explanation:
+        'Cannot start--this collector is already started.',
+      });
+    } else if (returnedCollector.status === 'Paused') {
+      throw new apiErrors.ForbiddenError({ explanation:
+        'Cannot start--this collector is paused. ' +
+        'Resume this collector instead.',
+      });
+    }
+
+    // doPatch needs the key in swagger params
+    req.swagger.params.key = { value: toPost.name };
+    return doPatch(req, res, next, helper, true);
```

Comment:
Can we not call `doPatch` here? I think we decided that for special cases like this we will do the job right in the specific controllers, rather than calling `doPatch` and adding `if` conditions there. 
In line 339 you already have the collector, can you just use that collector instance and update it.